helix-term/src/commands/typed.rs

```diff
@@ -2408,6 +2409,82 @@ fn redraw(
     Ok(())
 }
 
+fn move_buffer(
+    cx: &mut compositor::Context,
+    args: &[Cow<str>],
+    event: PromptEvent,
+) -> anyhow::Result<()> {
+    if event != PromptEvent::Validate {
+        return Ok(());
+    }
+
+    ensure!(args.len() == 1, format!(":move takes one argument"));
+    let doc = doc!(cx.editor);
+
+    let new_path_as_url = Url::from_file_path(get_canonicalized_path(&PathBuf::from(
+        args.first().unwrap().to_string(),
+    )))
+    .unwrap();
+    let old_path_as_url = doc
+        .url()
+        .ok_or_else(|| anyhow!("Scratch buffer cannot be moved. Use :write instead"))?;
+
+    let edits: Vec<(
+        helix_lsp::Result<helix_lsp::lsp::WorkspaceEdit>,
+        OffsetEncoding,
+        String,
+    )> = doc
+        .language_servers()
+        .map(|lsp| {
+            (
+                lsp.prepare_file_rename(&old_path_as_url, &new_path_as_url),
+                lsp.offset_encoding(),
+                lsp.name().to_owned(),
+            )
+        })
+        .filter(|(f, _, _)| f.is_some())
+        .map(|(f, encoding, name)| (helix_lsp::block_on(f.unwrap()), encoding, name))
+        .collect();
+
+    for (lsp_reply, encoding, name) in edits {
+        match lsp_reply {
+            Ok(edit) => {
+                if let Err(e) = apply_workspace_edit(cx.editor, encoding, &edit) {
+                    log::error!(
+                        ":move command failed to apply edits from lsp {}: {:?}",
+                        name,
+                        e
+                    );
+                };
+            }
+            Err(e) => {
+                log::error!("LSP {} failed to treat willRename request: {:?}", name, e);
+            }
+        };
+    }
+
+    let doc = doc_mut!(cx.editor);
+    let old_path = old_path_as_url.to_file_path().unwrap();
+    let new_path = new_path_as_url.to_file_path().unwrap();
+
+    doc.set_path(Some(new_path.as_path()));
+    if let Err(e) = std::fs::rename(&old_path, &new_path) {
+        doc.set_path(Some(old_path.as_path()));
+        bail!("Could not move file: {}", e);
+    };
+
+    doc.language_servers().for_each(|lsp| {
+        lsp.did_file_rename(&old_path_as_url, &new_path_as_url);
+    });
```

Comment:
> I think the Document path should only be set if the file rename succeeds.

Hmm that's the case already, isn't it ? There's an error returned if the call to `std::fs:rename` fails.

Reply:
while the current code is technically correct it would be clearer/more correct to just set the path once after the rename succeded.

Reply:
> Also, when I was testing my own version of the rename feature, rust-analyzer seemed to accept the rename even without DidRenameFiles, so I'm not sure what would happen if the file rename failed

we probably have to rollback the edit if the rename fails. we currently don't implement that but we will want transactional lsp edits eventually which will allow that.

Reply:
> while the current code is technically correct it would be clearer/more correct to just set the path once after the rename succeded instead of setting it immidiedly and then changing it back.

Oh ok, we are talking about the `doc.set_path`, sorry :sweat_smile: 

I can do that, I only implemented in that way following [that](https://github.com/helix-editor/helix/pull/5531/files#r1255790329) comment on the previous PR :smile: 

But I do tend to agree with that comment though. Meaning that if the rename on the FS goes through, but not for the doc in helix, it's indeed a bit more touchy to rollback the change on the FS.

Reply:
Yeah the rollbacks are a bit of a weird situation, I couldn't find anything about failing to rename a file on the LSP docs, or anywhere actually. I think rolling back with both WillRenameFiles and DidRenameFiles is the best option.

As for `doc.set_path`, I think that's something that can be relied upon, since there shouldn't be a reason that setting the `Document`'s path should fail, other than a bug with the method itself

Reply:
I forgot to mention, if I remember correctly, rust-analyzer doesn't even recognize DidRenameFiles, though my memory is a bit foggy there, I'll get around to testing that tmr and I'll drop what I find here

Reply:
yeah rust-analyzer does not have the didRenameFile capability.
WillReanmeFile does not affect any RA state tough only once we actually send the corresponding transcation to the server (which is why we could rollback easily). 


The whole rollback thing is a can of worms I don't want to open in this PR that is an existing issue

Reply:
well maybe rust-analyser don't care about it but other lsp might ^^

I don't think though we should rollback things on the lsp if anything bad happens. Probably the edits but I haven't been able to reproduce a case where a LSP provides edit to apply while having the rename on filesystem to fail.

Somehow I think rust-analyzer is able to forecast that we won't be able to rename the file as we wish to

Reply:
> Somehow I think rust-analyzer is able to forecast that we won't be able to rename the file as we wish to

rust-analyzer can't forecast that. IO is inherently non-deterministic. You are right its not a huge issue in practice but it definitely is in edgecases and  vscode does implement that. Ubt like I said rolling back the edits is a bit hard and doesn't need to be done in this PR.